src/Symfony/Component/Security/Http/Logout/LogoutUrlGenerator.php

```diff
@@ -86,24 +98,31 @@ public function getLogoutUrl($key = null)
      */
     private function generateLogoutUrl($key, $referenceType)
     {
-        // Fetch the current provider key from token, if possible
-        if (null === $key && null !== $this->tokenStorage) {
-            $token = $this->tokenStorage->getToken();
-            if (null !== $token && method_exists($token, 'getProviderKey')) {
+        $guess = $key === null;
+
+        if ($guess && null !== $this->tokenStorage && null !== $token = $this->tokenStorage->getToken()) {
+            // There is no logout url to generate for an anonymous token
+            if ($token instanceof AnonymousToken) {
+                return;
+            }
+
+            // Fetch the current provider key from token, if possible
+            if (method_exists($token, 'getProviderKey')) {
                 $key = $token->getProviderKey();
             }
         }
 
-        if (null === $key) {
+        if (null === $key && null === $this->currentFirewall) {
```

Comment:
I prefer not: I'd rather keep the `$key` argument mandatory for calling `getListener()`